### PR TITLE
CLOUDSTACK-9184: Fixes #2631 VMware dvs portgroup autogrowth

### LINKED ISSUE #2631: New vmware.ports.per.dvportgroup global setting breaks existing deployments
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
Hypervisor (VMware)
~~~

##### CLOUDSTACK VERSION

~~~
4.11.0
~~~

##### CONFIGURATION

dvSwitch in VMware environment >5.0

##### SUMMARY

~~~
commit:
https://github.com/sureshanaparti/cloudstack/commit/deb2f530b71c1dc13665c0b8dacfb025c38f58b9#diff-2e10685e0f701ef7fd43267f2879bb8e

sets vmware.ports.per.dvportgroup to a low value allowing autogrow to set the value going forward.  However, If a portgroup already exists and has more than (say) 10 ports in use already then a error such as:

> Cannot set number of ports to 10 because at least 11 out of 18 ports are still in use

will be thrown when creating a VM on said dvSwitch and the VM will fail to be deployed.
Setting ALL dvsportgroups to use a sufficiently high number to cover ALL dvsportgroup values will be expensive in large environments.


~~~

##### STEPS TO REPRODUCE

~~~
Create a VMware environment >5.0 (5.5 in this case), which uses dvSwitches - this example manifests itself on the mgmt network.  CloudStack version 4.6
use the environment sufficiently to create more than 10 ports on mgmt dvSwitch (or any other network to create other failures).

Upgrade to ACS 4.11.0

Attempt to destroy CPVM in order for it to be recreated.
~~~

##### EXPECTED RESULTS

~~~
CPVM is destroyed and recreated (CloudStack either leaves number of ports attached 'as is' or reduces the number of ports in the portgroup to slightly greater than the number of ports in use.
~~~

##### ACTUAL RESULTS

~~~
An error is thrown:

> Cannot set number of ports to 10 because at least 11 out of 18 ports are still in use

CPVM is not deployed.
~~~

### PR DESCRIPTION
**Pull request by rhtyd**
This deprecates the vmware.ports.per.dvportgroup global setting.

The vSphere Auto Expand feature (introduced in vSphere 5.0) will take
care of dynamically increasing/decreasing the dvPorts when running out
of distributed ports . But in case of vSphere 4.1/4.0 (If used), as this
feature is not there, the new default value (=> 8) have an impact in the
existing deployments. Action item for vSphere 4.1/4.0: Admin should
modify the global configuration setting "vmware.ports.per.dvportgroup"
from 8 to any number based on their environment because the proposal
default value of 8 would be very less without auto expand feature in
general. The current default value of 256 may not need immediate
modification after deployment, but 8 would be very less which means
admin need to update immediately after upgrade.

This continues and take-over from PR https://github.com/apache/cloudstack/pull/1253

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

I deployed ACS 4.6 and performed an upgrade to latest 4.11 with dvs, I could not reproduce the issues shared on #2631 

## Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
Testing
- [ ] I have added tests to cover my changes.
- [x] All relevant new and existing integration tests have passed.
- [x] A full integration testsuite with all test that can run on my environment has passed.

### CODE REVIEW

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-65

**Comment by rhtyd** [comment]
@sureshanaparti I've concluded a standalone and upgrade test (upgrade from 4.6) and conclude no regressions. Do you have any comments on the PR?

File: plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/manager/VmwareManagerImpl.java [hunk 1]
```diff
@@ -394,8 +394,16 @@ private void prepareHost(HostMO hostMo, String privateTrafficLabel) throws Excep
             HypervisorHostHelper.prepareNetwork(vSwitchName, "cloud.private", hostMo, vlanId, null, null, 180000, false, BroadcastDomainType.Vlan, null, null);
         }
         else {
+            int portsPerDvPortGroup = _portsPerDvPortGroup;
+            AboutInfo about = hostMo.getHostAboutInfo();
+            if (about != null) {
+                String version = about.getApiVersion();
+                if (version != null && (version.equals("4.0") || version.equals("4.1")) && _portsPerDvPortGroup < DEFAULT_PORTS_PER_DV_PORT_GROUP_VSPHERE4_x) {
```
Comment (rafaelweingartner) [inline plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/manager/VmwareManagerImpl.java]: do you need that second condition there `_portsPerDvPortGroup < DEFAULT_PORTS_PER_DV_PORT_GROUP_VSPHERE4_x`?

File: plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/manager/VmwareManagerImpl.java [hunk 2]
```diff
@@ -394,8 +394,16 @@ private void prepareHost(HostMO hostMo, String privateTrafficLabel) throws Excep
             HypervisorHostHelper.prepareNetwork(vSwitchName, "cloud.private", hostMo, vlanId, null, null, 180000, false, BroadcastDomainType.Vlan, null, null);
         }
         else {
+            int portsPerDvPortGroup = _portsPerDvPortGroup;
+            AboutInfo about = hostMo.getHostAboutInfo();
+            if (about != null) {
+                String version = about.getApiVersion();
```
Comment (rafaelweingartner) [inline plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/manager/VmwareManagerImpl.java]: Are the versions fixed? I mean, can't we have `4.2`, `4.3`, and so on.

Comment (sureshanaparti) [reply plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/manager/VmwareManagerImpl.java]: @rafaelweingartner No VMware ESXi/ESX hosts with these release versions. Take a look at the versions in the article https://kb.vmware.com/s/article/2143832.

Comment (rhtyd) [reply plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/manager/VmwareManagerImpl.java]: LGTM

**Comment by rhtyd** [comment]
Test LGTM. Are you lgtm on this @rafaelweingartner ?

**Comment by rhtyd** [comment]
Note: While I sent the PR, I'll give my LGTM as the changes are not authored by me.

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-65

**Comment by rhtyd** [comment]
Manually tested, also verified on trillian env. Based on review and trillian tests results (including incoming/2nd run), merging this.

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-65

**Pull request merged**